File: Dynamic programming/313_super_ugly_number.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector<int> dp(n);
        int m = primes.size();
        vector<int> point(m);

        for(int i=1; i<n; ++i) {
            dp[i] = dp[point[0]] * primes[0];
            for(int j=1; j<m; ++j) {
                dp[i] = min(dp[i], dp[point[j]] * primes[j]);
            }
            for(int j=0; j<m; ++j) {
                if (dp[i] == dp[point[j]] * primes[j]) point[j] ++;
            }
        }
        return dp[n-1];
    }
};

File: Dynamic programming/313_super_ugly_number.py
class Solution:
    def nthSuperUglyNumber(self, n: int, primes: List[int]) -> int:
        # simillar with #264
        # dynamic programing
        dp = [0] * n
        m = len(primes)
        point = [0] * m
        
        for i in range(1, n):
            dp[i] = min(dp[point[j]] * primes[j] for j in range(m))
            for j in range(m):
                if dp[i] == dp[point[j]] * primes[j]:
                    point[j] += 1
        return dp[n-1]
